File: web-app/backend/app/api/v1/endpoints/dhasa.py
"""
Dhasa calculation endpoints
All 47 types of dhasa systems
"""
from fastapi import APIRouter, HTTPException, Query
from app.models.schemas import DhasaRequest, ErrorResponse
from app.services.dhasa_service import DhasaService
from typing import Dict, Any

router = APIRouter()
dhasa_service = DhasaService()


@router.post("/vimsottari", response_model=Dict[str, Any])
async def get_vimsottari_dhasa(request: DhasaRequest):
    """
    Calculate Vimsottari Dhasa (most popular)
    120-year cycle based on Moon's nakshatra
    """
    try:
        result = dhasa_service.get_vimsottari_dhasa(
            request.birth_details.dict(),
            request.include_antardhasa,
            request.ayanamsa,
            request.max_sub_level or 2,
            request.focus_mahadasha_index
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/graha/{dhasa_type}", response_model=Dict[str, Any])
async def get_graha_dhasa(dhasa_type: str, request: DhasaRequest):
    """
    Calculate any Graha/Nakshatra based dhasa:
    - vimsottari, ashtottari, yogini, shodasottari
    - dwadasottari, dwisapathi, panchottari, sataatbika
    - chathuraaseethi_sama, shastihayani, shattrimsa_sama
    - naisargika, tara, karaka, aayu
    """
    try:
        result = dhasa_service.get_any_graha_dhasa(
            request.birth_details.dict(),
            dhasa_type,
            request.include_antardhasa,
            request.ayanamsa,
            request.max_sub_level or 2,
            request.focus_mahadasha_index
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/raasi/{dhasa_type}", response_model=Dict[str, Any])
async def get_raasi_dhasa(dhasa_type: str, request: DhasaRequest):
    """
    Calculate any Raasi based dhasa:
    - narayana, kendradhi_rasi, sudasa, drig, nirayana
    - shoola, chara, lagnamsaka, padhanadhamsa, mandooka
    - sthira, tara_lagna, brahma, varnada, yogardha
    - navamsa, paryaaya, trikona, kalachakra, moola, chakra
    """
    try:
        result = dhasa_service.get_any_raasi_dhasa(
            request.birth_details.dict(),
            dhasa_type,
            request.include_antardhasa,
            request.ayanamsa,
            request.max_sub_level or 2
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/applicable", response_model=Dict[str, Any])
async def get_applicable_dhasas(request: DhasaRequest):
    """
    Get list of all applicable dhasa systems for the chart
    """
    try:
        result = dhasa_service.get_all_applicable_dhasas(
            request.birth_details.dict(),
            request.ayanamsa
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: web-app/frontend/lib/api/dhasa.ts
import { apiClient } from './client';
import { BirthDetails } from './charts';

export interface DhasaRequest {
  birth_details: BirthDetails;
  dhasa_type: string;
  include_antardhasa?: boolean;
  ayanamsa?: string;
  max_sub_level?: number;
  focus_mahadasha_index?: number;
}

export interface DhasaPeriod {
  planet: string; // or rasi
  planet_index?: number;
  start_date: string;
  end_date: string;
  duration_years?: number;
  sub_periods?: DhasaPeriod[];
}

export interface DhasaResponse {
  birth_details: BirthDetails;
  dhasa_type: string;
  balance_at_birth: any;
  periods: DhasaPeriod[];
}

const GRAHA_DHASAS = [
  'vimsottari', 'ashtottari', 'yogini', 'shodasottari', 'dwadasottari',
  'dwisatpathi', 'panchottari', 'sataatbika', 'chathuraaseethi_sama',
  'shastihayani', 'shattrimsa_sama', 'naisargika', 'tara', 'karaka', 'aayu'
];

const RAASI_DHASAS = [
  'narayana', 'kendradhi_rasi', 'sudasa', 'drig', 'nirayana', 'shoola',
  'chara', 'lagnamsaka', 'padhanadhamsa', 'mandooka', 'sthira', 'tara_lagna',
  'brahma', 'varnada', 'yogardha', 'navamsa', 'paryaaya', 'trikona',
  'kalachakra', 'moola', 'chakra'
];

export const dhasaApi = {
  calculateVimsottari: async (data: Omit<DhasaRequest, 'dhasa_type'>): Promise<DhasaResponse> => {
    const response = await apiClient.post<DhasaResponse>('/api/v1/dhasa/vimsottari', {
      ...data,
      dhasa_type: 'vimsottari'
    });
    return response.data;
  },

  calculateDhasa: async (data: DhasaRequest): Promise<DhasaResponse> => {
    const { dhasa_type, ...rest } = data;
    let category = 'graha';

    if (GRAHA_DHASAS.includes(dhasa_type)) {
      category = 'graha';
    } else if (RAASI_DHASAS.includes(dhasa_type)) {
      category = 'raasi';
    } else {
      // Default or Annual fallback
      category = 'graha';
    }

    const response = await apiClient.post<DhasaResponse>(`/api/v1/dhasa/${category}/${dhasa_type}`, data);
    return response.data;
  },

  getApplicableDhasas: async (data: Omit<DhasaRequest, 'dhasa_type'>) => {
    const response = await apiClient.post('/api/v1/dhasa/applicable', data);
    return response.data;
  },
};
